gocached/gocached.go

```diff
@@ -0,0 +1,1349 @@
+	db.SetConnMaxLifetime(0) // no limit
+	if _, err := db.Exec(schema); err != nil {
+		return nil, err
+	}
+	return db, nil
+}
+
+// Start initializes the server. Must be called before serving requests.
+func (srv *Server) Start(ctx context.Context) error {
+	if srv.Logf == nil {
+		srv.Logf = log.Printf
+	}
+	if err := os.MkdirAll(srv.Dir, 0750); err != nil {
+		return fmt.Errorf("creating cache dir: %w", err)
+	}
+	db, err := openDB(srv.Dir)
+	if err != nil {
+		return fmt.Errorf("openDB: %w", err)
+	}
+	srv.db = db
+	srv.sessions = make(map[string]*sessionData)
+	srv.shutdownCtx, srv.shutdownCancel = context.WithCancel(ctx)
+
+	reg := prometheus.NewRegistry()
+	reg.MustRegister(
+		collectors.NewGoCollector(),
+		collectors.NewProcessCollector(collectors.ProcessCollectorOpts{}),
+		collectors.NewBuildInfoCollector(),
+	)
+	srv.registerMetrics(reg)
+
+	srv.metricsHandler = promhttp.HandlerFor(reg, promhttp.HandlerOpts{
+		ErrorLog: log.Default(),
+	})
+
+	log.Printf("gocached: scanning usage & cleaning as needed...")
+	us, err := srv.usageStats()
+	if err != nil {
+		return fmt.Errorf("getting usage stats: %w", err)
+	}
+
+	log.Printf("gocached: current usage: %v of limit %v", us.All(), bytesFmt(srv.MaxSize))
+	if res, err := srv.cleanOldObjects(us); err != nil {
+		return fmt.Errorf("clean old objects: %w", err)
+	} else if res.Count > 0 {
+		log.Printf("gocached: cleaned %v", res)
+	}
+
+	if srv.JWTIssuer != "" {
+		srv.jwtValidator = ijwt.NewJWTValidator(srv.JWTIssuer, gocachedAudience)
+		if err := srv.jwtValidator.RunUpdateJWKSLoop(srv.shutdownCtx); err != nil {
+			return fmt.Errorf("failed to fetch JWKS for JWT validator: %w", err)
+		}
+
+		log.Printf("gocached: using JWT issuer %q with claims %v, global claims %v", srv.JWTIssuer, srv.JWTClaims, srv.GlobalJWTClaims)
+
+		go srv.runCleanSessionsLoop()
+	}
+
+	go srv.runCleanLoop()
+
+	return nil
+}
+
+func (s *Server) registerMetrics(reg *prometheus.Registry) {
+	rv := reflect.ValueOf(s).Elem()
+	t := reflect.TypeOf(s).Elem()
+	for i := 0; i < t.NumField(); i++ {
+		sf := t.Field(i)
+		if sf.Type == reflect.TypeFor[expvar.Int]() {
+			expvarInt := rv.Field(i).Addr().Interface().(*expvar.Int)
+			typ := sf.Tag.Get("type")
+			name := sf.Tag.Get("name")
+			if typ == "" {
+				panic("missing type tag for " + sf.Name)
+			}
+			if name == "" {
+				panic("missing name tag for " + sf.Name)
+			}
+			help := sf.Tag.Get("help")
+			metricName := "gocached_" + name
+
+			if tag := sf.Tag.Get("type"); tag != "" {
+				if tag == "gauge" {
+					reg.MustRegister(singleMetricCollector{&expvarGaugeMetric{
+						desc: prometheus.NewDesc(metricName, help, nil, nil),
+						v:    expvarInt,
+					}})
+				} else if tag == "counter" {
+					reg.MustRegister(singleMetricCollector{&expvarCounterMetric{
+						desc: prometheus.NewDesc(metricName, help, nil, nil),
+						v:    expvarInt,
+					}})
+				}
+			}
+		}
+	}
+}
+
+// Server is a gocached server. It implements [http.Handler]. Must call
```

Comment:
"Must call ..." should be a complete sentence if we're acting like Go doc style.

But that part's kinda awkward API-wise. Maybe we should do Start implicitly with a sync.Once (and context.Background) if it's never Started by hand?

Reply:
I went back and forth on this one a bit because `Start` needs a context but I didn't want to expose a public `Ctx` field in `Server` (it has to be a long-lived context separate from any per-request JWT validation contexts), and there are too many options for a simple `NewServer(...` method, but I also don't hugely like the `ServerOpts` struct pattern.

Maybe a functional option style `NewServer(opts ...ServerOption)`? And then all the `Server` fields can be private, and users can optionally provide a `gocached.WithShutdownContext(ctx)` opt.